pkg/gce-pd-csi-driver/controller.go

```diff
@@ -212,7 +212,11 @@ func (gceCS *GCEControllerServer) CreateVolume(ctx context.Context, req *csi.Cre
 
 	volumeID, err := common.KeyToVolumeID(volKey, gceCS.CloudProvider.GetDefaultProject())
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Failed to convert volume key to volume ID: %v", err)
+		klog.Errorf("Failed to convert volume key to volume ID: %w", err)
```

Comment:
Ah, so much boilerplate for every error. It will be hard to be consistent about all of this in the future. Instead of IsUserError could we have something like CodeForError so the whole thing looks like the following?

if err != nil {
  return nil, status.Errorf(CodeForError(err), "Failed to convert volume key to volume ID: %v", err)
}

That does mean that we don't get the klog---but we weren't getting that before anyway. So maybe we could have
if err != nil {
  return nil, LoggedError("Failed to convert volume key to volume ID: %w", err)
}

and then LoggedError will klog as well as produce the status.Errorf.

Reply:
golang error handling is crap as far as I can tell---in addition to all the boilerplate, go's aversion to stack traces mean you never really know where an error originally came from.

In particular I like doing a klog in addition to an error return, because the klog gives you generally better information about where the error came from---but the error return is necessary as well because that will get plumbed into an RPC reply which will give clues in the sidecars or kubelet logs.

Maybe if LoggedError(msg, err) looks like { klog(msg, err); return status.Errorf(code, err)}, and we use %v in msg, then it will be enough? We don't care about the %v conversion in the klog, as we only human-read that, it's the status.Errorf call that's important?

Reply:
In general, if we are passing the error as a return value, I prefer to leave it up to the caller to log and only log if we are going to swallow an error and not pass it up. That helps avoid duplicate logging of errors at multiple levels.

Reply:
Generally I've had more problems with seeing an error up high in the stack, and not being able to trace back to where it was emitted, than any ambiguities around duplicated messages in the logs. That may be the style in go but if it's possible to push back against it, I'd like to....